File: rsleetcode/src/easy/convert_sorted_array_to_binary_search_tree.rs
//! 108. Convert Sorted Array to Binary Search Tree
//!
//! https://leetcode.com/problems/convert-sorted-array-to-binary-search-tree/

use std::cell::{Ref, RefCell};
use std::rc::Rc;
type Tree = Rc<RefCell<TreeNode>>;
type OptionTree = Option<Rc<RefCell<TreeNode>>>;

// Definition for a binary tree node.
#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

pub struct Solution {}

impl Solution {
    pub fn sorted_array_to_bst(nums: Vec<i32>) -> Option<Rc<RefCell<TreeNode>>> {
        fn create(l: usize, r: usize, nums: &Vec<i32>) -> OptionTree {
            if nums.is_empty() || l >= r {
                return None;
            };
            let m = (l + r) / 2;
            let mut node = TreeNode::new(nums[m]);

            node.left = create(l, m, nums);
            node.right = create(m + 1, r, nums);

            Some(Rc::new(RefCell::new(node)))
        }

        create(0, nums.len(), &nums)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn to_tree(nums: Vec<Option<i32>>) -> OptionTree {
        fn create_node(i: usize, nums: &Vec<Option<i32>>) -> OptionTree {
            if i >= nums.len() {
                return None;
            };
            if let Some(val) = nums.get(i)? {
                let mut node = TreeNode::new(*val);
                node.left = create_node(i * 2 + 1, &nums);
                node.right = create_node(i * 2 + 2, &nums);
                Some(Rc::new(RefCell::new(node)))
            } else {
                return None
            }
        }
        create_node(0usize, &nums)
    }

    #[test]
    fn test_1() {
        let res = vec![Some(0), Some(-3), Some(9), Some(-10), None, Some(5)];
        let root: OptionTree = Solution::sorted_array_to_bst(vec![-10, -3, 0, 5, 9]);
        assert_eq!(to_tree(res), root);
    }

    #[test]
    fn test_2() {
        let res = vec![Some(3), Some(1)];
        let root: OptionTree = Solution::sorted_array_to_bst(vec![1, 3]);
        assert_eq!(to_tree(res), root);
    }
}

File: leetcode/pyleetcode/pyleetcode/easy/convert_sorted_array_to_binary_search_tree.py
"""108. Convert Sorted Array to Binary Search Tree

https://leetcode.com/problems/convert-sorted-array-to-binary-search-tree/
"""

from __future__ import annotations

import enum
from typing import Optional, List  # noqa: F401


class TreeNode:

    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

    def __eq__(self, other: TreeNode) -> bool:
        s_stack = [self]
        o_stack = [other]
        while s_stack and o_stack:
            s_node = s_stack.pop()
            o_node = o_stack.pop()
            if s_node is None and o_node is None:
                continue

            if s_node is None and o_node is not None:
                return False
            if s_node is not None and o_node is None:
                return False
            if s_node.val != o_node.val:
                return False

            s_stack.extend([s_node.left, s_node.right])
            o_stack.extend([o_node.left, o_node.right])
        return True

    @staticmethod
    def from_list(nums: list[int]) -> TreeNode:
        return Solution.create_tree(nums)

    def __str__(self) -> str:
        return str(self.val)

    def __repr__(self) -> str:
        return str(self)


class D(enum.Enum):
    LEFT = 0
    RIGHT = 1


class Solution:

    def sortedArrayToBST(self, nums: List[int]) -> Optional[TreeNode]:
        if not nums:
            return None
        l = 0
        r = len(nums)
        m = (l + r) // 2
        root = TreeNode(nums[m])
        res = root
        stack = [(l, m, D.LEFT, root), (m + 1, r, D.RIGHT, root)]
        while stack:
            l, r, d, root = stack.pop()
            if l >= r:
                continue
            m = (l + r) // 2
            node = TreeNode(nums[m])
            if d is D.LEFT:
                root.left = node
            else:
                root.right = node

            stack.extend([(l, m, D.LEFT, node), (m + 1, r, D.RIGHT, node)])

        return res

    def process(self, *args, **kwargs) -> Optional[TreeNode]:  # noqa: N802
        return self.sortedArrayToBST(*args, **kwargs)

    @staticmethod
    def create_tree(nums: List[int]) -> Optional[TreeNode]:

        def create(i, vals):
            if i >= len(vals) or vals[i] is None:
                return None
            node = TreeNode(vals[i])
            node.left = create(i * 2 + 1, vals)
            node.right = create(i * 2 + 2, vals)
            return node

        return create(0, nums)
